crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+use std::{collections::VecDeque, ops::Range};
+
+use itertools::Itertools as _;
+
+use re_types_core::SizeBytes;
+
+// ---
+
+/// A [`FlatVecDeque`] that can be erased into a trait object.
+///
+/// Methods that don't require monomorphization over `T` are made dynamically dispatchable.
+pub trait ErasedFlatVecDeque: std::any::Any {
+    fn as_any(&self) -> &dyn std::any::Any;
+
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any;
+
+    fn into_any(self: Box<Self>) -> Box<dyn std::any::Any>;
+
+    /// Dynamically dispatches to [`FlatVecDeque::remove`].
+    ///
+    /// This is prefixed with `dyn_` to avoid method dispatch ambiguities that are very hard to
+    /// avoid even with explicit syntax and that silently lead to infinite recursions.
+    fn dyn_remove(&mut self, at: usize);
+
+    /// Dynamically dispatches to [`FlatVecDeque::remove`].
+    ///
+    /// This is prefixed with `dyn_` to avoid method dispatch ambiguities that are very hard to
+    /// avoid even with explicit syntax and that silently lead to infinite recursions.
+    fn dyn_remove_range(&mut self, range: Range<usize>);
+
+    /// Dynamically dispatches to [`FlatVecDeque::truncate`].
+    ///
+    /// This is prefixed with `dyn_` to avoid method dispatch ambiguities that are very hard to
+    /// avoid even with explicit syntax and that silently lead to infinite recursions.
+    fn dyn_truncate(&mut self, at: usize);
+}
+
+impl<T: 'static> ErasedFlatVecDeque for FlatVecDeque<T> {
+    #[inline]
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    #[inline]
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+        self
+    }
+
+    #[inline]
+    fn into_any(self: Box<Self>) -> Box<dyn std::any::Any> {
+        self
+    }
+
+    #[inline]
+    fn dyn_remove(&mut self, at: usize) {
+        FlatVecDeque::<T>::remove(self, at);
+    }
+
+    #[inline]
+    fn dyn_remove_range(&mut self, range: Range<usize>) {
+        FlatVecDeque::<T>::remove_range(self, range);
+    }
+
+    #[inline]
+    fn dyn_truncate(&mut self, at: usize) {
+        FlatVecDeque::<T>::truncate(self, at);
+    }
+}
+
+// ---
+
+/// A double-ended queue implemented with a pair of growable ring buffers, where every single
+/// entry is a flattened array of values.
+///
+/// You can think of this as the native/deserialized version of an Arrow `ListArray`.
+///
+/// This is particularly when working with many small arrays of data (e.g. Rerun's `TimeSeriesScalar`s).
+//
+// TODO(cmc): We could even use a bitmap for T=Option<Something>, which would bring this that much
+// closer to a deserialized version of an Arrow array.
+#[derive(Debug, Clone)]
+pub struct FlatVecDeque<T> {
+    values: VecDeque<T>,
+    offsets: VecDeque<usize>,
```

Comment:
needs a docstring. Is there always a zero at the start? Is the length of it the same as the number of logical vectors in the dequeue?